### ISSUE TITLE
FYI: helm-hl-todo-items

### ISSUE THREAD

**Issue by alphapapa** [emacs-helm/helm#1834]
Hi Thierry,

I just added this to the wiki as an example:

```el
(defun helm-hl-todo-items ()
    "Show `hl-todo'-keyword items in buffer."
    (helm :sources (helm-build-in-buffer-source "hl-todo items"
                     :init (lambda ()
                             (with-current-buffer (helm-candidate-buffer 'global)
                               (goto-char (point-min))
                               (delete-non-matching-lines hl-todo-regexp)))
                     :get-line #'buffer-substring)))
```

It was hard for me to figure out how to make it work, to make it pre-filter the lines, so I was glad that I finally was able to, and I thought the example might help others.

@tarsius I don't know if you would want to use `with-eval-after-load` to add this to `hl-todo` for Helm users, but I would be glad if you did.  :)

**Comment by tarsius** [comment]
@alphapapa Please open a pr so that we don't forget about it. I will think about it later. Little busy right now.

**Comment by thierryvolpiatto** [comment]
alphapapa <notifications@github.com> writes:

> Hi Thierry,
>
> I just added this to the wiki as an example:
>
> (defun helm-hl-todo-items ()
>     "Show `hl-todo'-keyword items in buffer."
>     (helm :sources (helm-build-in-buffer-source "hl-todo items"
>                      :init (lambda ()
>                              (with-current-buffer (helm-candidate-buffer 'global)
>                                (goto-char (point-min))
>                                (delete-non-matching-lines hl-todo-regexp)))
>                      :get-line #'buffer-substring)))

Thanks, but I wonder how this could work without feeding the candidate
buffer with something.

-- 
Thierry
Gpg Key fingerprint = 6CEC 7081 AB33 E251 4AB8  5FC2 28D1 7F53 59F2 9997

**Comment by alphapapa** [comment]
> Thanks, but I wonder how this could work without feeding the candidate
buffer with something.

Me too!  :)  I read the docstring for `helm-source-in-buffer` and grepped for usage of `helm-build-in-buffer-source` and googled for it and through lots of trial-and-error I finally found this combination that worked.

BTW, the docstring for `helm-source-in-buffer` says that `:match` can be a list of functions, but when I tried passing it a list with my own lambda in it, I got an assertion error that seemed to indicate that `identity` is the only acceptable function.  I tried adding it to the list after my lambda, but I got the same error.

**Comment by thierryvolpiatto** [comment]
alphapapa <notifications@github.com> writes:

>     Thanks, but I wonder how this could work without feeding the candidate
>     buffer with something.
>
> Me too! :) I read the docstring for helm-source-in-buffer and grepped
> for usage of helm-build-in-buffer-source and googled for it and
> through lots of trial-and-error I finally found this combination that
> worked.

I guess your buffer have been feeded by the previous attempts to make
something working, if you delete the *helm candidates:<something>*
buffer or restart emacs you will see it is not working, well you get an
empty buffer, what you want is something like this assuming you want to
search in current-buffer:

    (defun helm-hl-todo-items ()
      "Show `hl-todo'-keyword items in buffer."
      (interactive)
      (helm :sources (helm-build-in-buffer-source "hl-todo items"
                       :init (lambda ()
                               (with-current-buffer (helm-candidate-buffer 'global)
                                 (insert (with-helm-current-buffer (buffer-string)))
                                 (goto-char (point-min))
                                 (delete-non-matching-lines hl-todo-regexp)))
                       :get-line #'buffer-substring)))

> BTW, the docstring for helm-source-in-buffer says that :match can be a
> list of functions, but when I tried passing it a list with my own
> lambda in it, I got an assertion error that seemed to indicate that
> identity is the only acceptable function.

Good my assertion is working fine :-)

> I tried adding it to the list after my lambda, but I got the same
> error.

Read on ALL the docstring:

    When using `candidates-in-buffer' its default value is `identity' and
    don't have to be changed, use the `search' slot instead.

to be more precise the value is '(identity).


Basically, when using in-buffer instead of sync sources the data is
contained in a buffer instead of a list and when you enter something in
minibuffer, helm search this buffer (helm-candidate-buffer) with
re-search-forward, it is why it is much faster than using string-match
on each element of a list.

--
Thierry
Gpg Key fingerprint = 6CEC 7081 AB33 E251 4AB8  5FC2 28D1 7F53 59F2 9997

**Comment by thierryvolpiatto** [comment]
I have corrected your example in wiki, closing now.

Thanks.

**Comment by c02y** [comment]
Hi, I received the following message when executing `helm-hl-todo-items` function.
~~~
Symbol’s value as variable is void: hl-todo-regexp
~~~

I use the following two functions, both of them will display this same error message.
~~~
(defun helm-hl-todo-items ()
      "Show `hl-todo'-keyword items in buffer."
      (interactive)
      (helm :sources (helm-build-in-buffer-source "hl-todo items"
                       :init (lambda ()
                               (with-current-buffer (helm-candidate-buffer 'global)
                                 (insert (with-helm-current-buffer (buffer-string)))
                                 (goto-char (point-min))
                                 (delete-non-matching-lines hl-todo-regexp)))
                       :get-line #'buffer-substring)))

;; or this one
 (defun helm-hl-todo-items ()
      "Show `hl-todo'-keyword items in buffer."
      (interactive)
      (helm :sources (helm-build-in-buffer-source "hl-todo items"
                       :data (current-buffer)
                       :candidate-transformer (lambda (candidates)
                                                (cl-loop for c in candidates
                                                         when (string-match hl-todo-regexp c)
                                                         collect c))
                       :get-line #'buffer-substring)
            :buffer "*helm hl-todo*"))
~~~

**Comment by alphapapa** [comment]
@c02y It appears that that variable no longer exists in the  `hl-todo` package.

**Comment by c02y** [comment]
@alphapapa Year, sorry about that I didn't check it first. 

According to the git log of hl-todo.el file, the `hl-todo-regexp` was rename to `hl-todo--regexp`, but I changed it in the function, and execute it, it failed too, showing
~~~
Wrong type argument: stringp, nil 
~~~

I believe that this part should be updated in wiki page then.

**Comment by alphapapa** [comment]
@c02y Why don't you enable `debug-on-error` and study the backtrace, then maybe you can fix it.

**Comment by MoHKale** [comment]
This function doesn't appear to be defined in helm, is there any intention of including it in the package or is it purely something you're expected to define in your own config.

Also, yeah, the example posted seems to have been broken due to changes in hl-todo. it's a small fix to update the changed variables name:

```elisp
(defun helm-hl-todo-items ()
  "Show `hl-todo'-keyword items in buffer."
  (interactive)
  (helm :sources (helm-build-in-buffer-source "hl-todo items"
                   :init (lambda ()
                           (with-current-buffer (helm-candidate-buffer 'global)
                             (insert (with-helm-current-buffer (buffer-string)))
                             (goto-char (point-min))
                             (delete-non-matching-lines hl-todo--regexp)))
                   :get-line #'buffer-substring)))
```

however, as mentioned by @c02y this also throws a type error when evaluated. The stacktrace is:

```
Debugger entered--Lisp error: (wrong-type-argument stringp nil)
  string-match("\\[:\\(upp\\|low\\)er:]" nil)
  isearch-no-upper-case-p(nil t)
  delete-non-matching-lines(nil)
  (save-current-buffer (set-buffer (helm-candidate-buffer 'global)) (insert (save-current-buffer (set-buffer (or (and (buffer-live-p helm-current-buffer) helm-current-buffer) (setq helm-current-buffer (current-buffer)))) (buffer-string))) (goto-char (point-min)) (delete-non-matching-lines hl-todo--regexp))
  **NOTE section removed because stacktrace is HUGE!!!**
  helm-hl-todo-items()
  funcall-interactively(helm-hl-todo-items)
  call-interactively(helm-hl-todo-items record nil)
  command-execute(helm-hl-todo-items record)
  helm-M-x(nil #("helm-hl-todo-items" 0 18 (match-part "helm-hl-todo-items")))
  funcall-interactively(helm-M-x nil #("helm-hl-todo-items" 0 18 (match-part "helm-hl-todo-items")))
  call-interactively(helm-M-x nil nil)
  command-execute(helm-M-x)
```.

**Comment by alphapapa** [comment]
That's probably because:

```el
(defvar-local hl-todo--regexp nil)
```

It's a buffer-local value with no global value, so in the `helm-candidate-buffer`, it is nil.  You'll need to get the value using `with-helm-current-buffer`.

**Comment by tarsius** [comment]
... and in that buffer the value of the variable may be nil too. To make sure it isn't additionally use the **function** `hl-todo--regexp`.
